### PR TITLE
Rake task to view email delivery attempts

### PR DESCRIPTION
**Pull request by ChrisBAshton**
We sometimes get queries from users complaining about emails they have not received.

To enable us to help them more easily, the `manage:view_emails` rake task attempts to pull together all email delivery attempts made to the given email address.

### CODE REVIEW

File: docs/support-tasks.md [hunk 1]
```diff
@@ -29,6 +29,19 @@ $ bundle exec rake manage:change_email_address[<old_email_address>, <new_email_a
 
 [change]: https://deploy.blue.production.govuk.digital/job/run-rake-task/parambuild/?TARGET_APPLICATION=email-alert-api&MACHINE_CLASS=email_alert_api&RAKE_TASK=manage:change_email_address[from@example.org,to@example.org]
 
+## View subscriber's recent emails
+
+This task shows the most recent email delivery attempts made to the given user.
+The history limit is 10 emails by default, but you can override this.
+
+```bash
+$ bundle exec rake manage:view_emails[<email_address>,<limit=10>]
```
[the remark this answers is not in the document]
Comment (ChrisBAshton) [reply docs/support-tasks.md:38]: Just `10` by itself is enough. I was trying to make it clear that the second parameter is optional and defaults to `10`, but I can see why this is confusing.

Would this be any clearer? It follows [PHP API convention](https://www.php.net/manual/en/function.preg-match.php) - I couldn't find a good example of optional params in the Ruby docs 😬 

```suggestion
$ bundle exec rake manage:view_emails[<email_address>[, int limit=10]]
```

File: docs/support-tasks.md [hunk 2]
```diff
@@ -29,6 +29,19 @@ $ bundle exec rake manage:change_email_address[<old_email_address>, <new_email_a
...
+```
+
+[⚙ Run rake task on production][view_emails]
+
+[view_emails]: https://deploy.blue.production.govuk.digital/job/run-rake-task/parambuild/?TARGET_APPLICATION=email-alert-api&MACHINE_CLASS=email_alert_api&RAKE_TASK=manage:view_emails[email@example.org,10]
```
Comment (ChrisBAshton) [inline docs/support-tasks.md:43]: Related - do you think we should omit the `,10` from the URL and rely on devs to add a second parameter if needed?

```suggestion
[view_emails]: https://deploy.blue.production.govuk.digital/job/run-rake-task/parambuild/?TARGET_APPLICATION=email-alert-api&MACHINE_CLASS=email_alert_api&RAKE_TASK=manage:view_emails[email@example.org]
```

File: docs/support-tasks.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (leenagupte) [reply docs/support-tasks.md:38]: I think this: 
`$ bundle exec rake manage:view_emails[<email_address>,<limit>]`

With an explanation of what the params represent would be better.

File: docs/support-tasks.md (hunk 2 above)
Comment (leenagupte) [reply docs/support-tasks.md:43]: Yeah. I think in most cases the default of 10 will suffice.

Comment (leenagupte) [reply docs/support-tasks.md:43]: It's nice to have the option of changing it though.

File: lib/tasks/manage.rake [hunk 3]
```diff
@@ -32,6 +32,46 @@ namespace :manage do
...
+
+    # Example output:
+    # | Status                  | SubscriberList            | Frequency | Timeline                                                       |
+    # | Inactive (unsubscribed) | Test foo (slug: test-foo) | daily     | Subscribed 2020-04-18 15:21:20 +0100, Ended 2020-05-12 11:41:04 +0100 |
+    # | Active                  | Bar bar (slug: bar-bar)   | weekly    | Subscribed 2019-06-10 13:48:43 +0100                           |
+    puts "| #{columns.map { |_, g| g[:label].ljust(g[:width]) }.join(' | ')} |"
+    results.each do |h|
+      str = h.keys.map { |k| h[k].ljust(columns[k][:width]) }.join(" | ")
+      puts "| #{str} |"
+    end
```
[the remark this answers is not in the document]
Comment (ChrisBAshton) [reply lib/tasks/manage.rake]: I've had a go at improving it.

File: docs/support-tasks.md [hunk 4]
```diff
@@ -29,6 +29,21 @@ $ bundle exec rake manage:change_email_address[<old_email_address>, <new_email_a
...
+`limit` defaults to 10, but you can override this if you need to see more of
+the user's history.
+
+```bash
+$ bundle exec rake manage:view_emails[<email_address>,<limit>]
+```
+
+[⚙ Run rake task on production][view_emails]
+
+[view_emails]: https://deploy.blue.production.govuk.digital/job/run-rake-task/parambuild/?TARGET_APPLICATION=email-alert-api&MACHINE_CLASS=email_alert_api&RAKE_TASK=manage:view_emails[email@example.org]
```
Comment (leenagupte) [inline docs/support-tasks.md:45]:
```suggestion
[view_emails]: https://deploy.blue.production.govuk.digital/job/run-rake-task/parambuild/?TARGET_APPLICATION=email-alert-api&MACHINE_CLASS=email_alert_api&RAKE_TASK=manage:view_emails["email@example.org"]
```

Nitpick: Should the email address be in quotes?

File: lib/tasks/manage.rake [hunk 5]
```diff
@@ -1,23 +1,33 @@
...
-    heading = col.to_s.humanize
-    h[col] = { label: heading, width: [hash.map { |g| g[col].size }.max, heading.size].max }
+  columns = []
+  hash.first.keys.each do |key|
+    heading = key.to_s.humanize
+    longest_value = hash.map { |row| row[key].to_s.size }.max
+    columns << {
+      id: key,
+      label: heading,
+      width: [longest_value, heading.size].max,
```
Comment (leenagupte) [inline lib/tasks/manage.rake:18]: Nice! This makes so much sense now! ⭐

File: spec/lib/tasks/manage_spec.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
+RSpec.describe "manage" do
+  describe "hash_to_table" do
+    it "converts a hash to table markdown" do
```
Comment (leenagupte) [inline spec/lib/tasks/manage_spec.rb:3]: Tests ❤️

File: docs/support-tasks.md (hunk 4 above)
Comment (ChrisBAshton) [reply docs/support-tasks.md:45]: It works without quotes, and the other rake task links don't have quotes in the URL either.

Thanks for the review :)

File: lib/tasks/manage.rake [hunk 7]
```diff
@@ -16,20 +46,33 @@ namespace :manage do
...
-    # | Inactive (unsubscribed) | Test foo (slug: test-foo) | daily     | Subscribed 2020-04-18 15:21:20 +0100, Ended 2020-05-12 11:41:04 +0100 |
-    # | Active                  | Bar bar (slug: bar-bar)   | weekly    | Subscribed 2019-06-10 13:48:43 +0100                           |
-    puts "| #{columns.map { |_, g| g[:label].ljust(g[:width]) }.join(' | ')} |"
-    results.each do |h|
-      str = h.keys.map { |k| h[k].ljust(columns[k][:width]) }.join(" | ")
-      puts "| #{str} |"
+  desc "View most recent email delivery attempts for a subscriber"
+  task :view_emails, %i[email_address limit] => :environment do |_t, args|
+    email_address = args[:email_address]
+    limit = args[:limit] || 10
```
Comment (brucebolt) [inline lib/tasks/manage.rake:55]: I think there might be an issue here.  The `args[:limit]` value will be a string, so this needs converting to an integer (see https://deploy.blue.production.govuk.digital/job/run-rake-task/78388/console).

**Pull request merged**